xapian-applications/omega/assistant.cc

```diff
@@ -0,0 +1,127 @@
+/** @file assistant.cc
+ * @brief Worker module for putting text extraction into a separate process.
+ */
+/* Copyright (C) 2011 Olly Betts
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+#include <csignal>
+
+#include "worker_comms.h"
+#include "handler_pdf.h"
+#include "safeunistd.h"
+
+using namespace std;
+
+const int FD = 3;
+const int time_limit = 300;
+
```

Comment:
I need to use an asynchronous timer to detect possible infinite loops. It is easy to use SIGALARM to implement it, but this is not going to work if we are not on a Unix system.
An alternative (much more expensive) is to use threads, but it is necessary create a new thread every time. So we are going to create as many threads as files for indexing.
Of course, it is possible just to avoid using it. However, without a timer we cannot detect infinite loops on a particular library.

Reply:
It seems reasonable to use `alarm()` and just not have such a timeout for platforms which don't provide that - it's specified by POSIX after all.  That makes for a simpler and lower overhead solution for most platforms, and indexing will still work on other platforms, just without a timeout.  People who care about other platforms can always contribute a patch to address this.

Another approach would be for the main omindex process to implement the timeout by using `poll()` or `select()` on the fd for the pipe to the subprocess.  Then if the timeout passes before it receives data it can just kill the subprocess (which should be a rare occurrence so the overhead of having to set up the subprocess again isn't a big concern).  Unfortunately Microsoft's socket implementation stupidly makes a socket a different thing to a file descriptor, and `select()` only works on sockets and not arbitrary file descriptors, so it'd probably still need special code (or to use sockets to communicate with the child process instead of a pipe).

In general I wouldn't worry too much about porting to Microsoft Windows for Omega code.  It's important that xapian-core works there as fully as possible, but I don't think I've heard of anyone using Omega on that platform recently.  So long as it builds and works (with slightly degraded functionality in some places), better to focus energy on getting more done on POSIX platforms.  You can always come back later and try to address the missing functionality.

Reply:
So, I just delete the threads option?

Reply:
I think so.  It's certainly good to be thinking about portability though.